raftlog: introduce & use in loqrecovery, kvserver

tbg opened the pull request:
This commit introduces a new package `raftlog`. Aspirationally, this
will at some point become the de facto way of encapsulating the raft log
encoding and may play a role in programmatically constructing and
inspecting raft logs (e.g.  for testing).

For now, we introduce two concepts:

- `raftlog.Entry`, which wraps a `raftpb.Entry` and all of the
  information derived from it, such as the command ID, the
`kvserverpb.RaftCommand`, the configuration change (if any), etc.

- `raftlog.Iterator`, a way to iterate over the raft log in terms of
  `raftlog.Entry` (as opposed to `raftpb.Entry` which requires lots of
manual processing).

Both are then applied across the codebase, concretely:

- `loqrecovery` is simplified via `raftlog.Iterator` to pull commit
  triggers out of the raft log.
- debug pretty-printing is simpler thanks to use of `raftlog.Entry`.
- `decodedRaftEntry` is now structurally a `raftpb.Entry`, and again
  lots manual custom unmarshaling code evaporates.

It's currently difficult to create "interesting" raft log entries if
trying to stay away from manual population of large datastructures
(which is prone to rotting), so there's zero unit testing of
`raftlog.Iterator`. However, the code is not new, instead it was
deduplicated from a few places, and is now tested through all of them;
so I don't feel to bad about it. I still think it is a priority to be
able to "comfortably" create at least "simple" raft logs, meaning we
need to be able to string together `batcheval` and entry creation at
least in a rudimentary fashion. I intend to look into this next and
add comprehensive unit tests for `raftlog.{Entry,Iterator}`.

Touches #75729.

Epic: CRDB-220
Release note: None

tbg committed:
raftlog: introduce & use in loqrecovery, kvserver

This commit introduces a new package `raftlog`. Aspirationally, this
will at some point become the de facto way of encapsulating the raft log
encoding and may play a role in programmatically constructing and
inspecting raft logs (e.g.  for testing).

For now, we introduce two concepts:

- `raftlog.Entry`, which wraps a `raftpb.Entry` and all of the
  information derived from it, such as the command ID, the
`kvserverpb.RaftCommand`, the configuration change (if any), etc.

- `raftlog.Iterator`, a way to iterate over the raft log in terms of
  `raftlog.Entry` (as opposed to `raftpb.Entry` which requires lots of
manual processing).

Both are then applied across the codebase, concretely:

- `loqrecovery` is simplified via `raftlog.Iterator` to pull commit
  triggers out of the raft log.
- debug pretty-printing is simpler thanks to use of `raftlog.Entry`.
- `decodedRaftEntry` is now structurally a `raftpb.Entry`, and again
  lots manual custom unmarshaling code evaporates.

It's currently difficult to create "interesting" raft log entries if
trying to stay away from manual population of large datastructures
(which is prone to rotting), so there's zero unit testing of
`raftlog.Iterator`. However, the code is not new, instead it was
deduplicated from a few places, and is now tested through all of them;
so I don't feel to bad about it. I still think it is a priority to be
able to "comfortably" create at least "simple" raft logs, meaning we
need to be able to string together `batcheval` and entry creation at
least in a rudimentary fashion. I intend to look into this next and
add comprehensive unit tests for `raftlog.{Entry,Iterator}`.

Touches #75729.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76126)
<!-- Reviewable:end -->

erikgrinaker commented on pkg/kv/kvserver/raftlog/entry.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Load populates the Entry from the Ent field. Will reset the Meta field.
+func (e *Entry) Load() error {
```
This should take a `raftpb.Entry` and place it in `Ent` instead of callers having to populate it themselves.

erikgrinaker commented on pkg/kv/kvserver/raftlog/entry.go:
```diff
@@ -0,0 +1,107 @@
+// Entry contains data related to a raft log entry. This is the raftpb.Entry
+// itself but also all encapsulated data relevant for command application.
+type Entry struct {
+	Meta enginepb.MVCCMetadata // only set after LoadFromRawValue
```
Can we get rid of this field? It doesn't appear to be used anywhere, it's unclear what purpose it serves, and it's unfortunate that it's sometimes uninitialized.

erikgrinaker commented on pkg/kv/kvserver/replica_application_cmd.go:
```diff
@@ -44,8 +42,7 @@ import (
 type replicatedCmd struct {
-	ent              *raftpb.Entry // the raft.Entry being applied
-	decodedRaftEntry               // decoded from ent
+	decodedRaftEntry // decoded from ent
```
Remove the comment -- `ent` no longer exists.

erikgrinaker commented on pkg/kv/kvserver/raftlog/entry.go:
```diff
@@ -0,0 +1,107 @@
+// itself but also all encapsulated data relevant for command application.
+type Entry struct {
+	Meta enginepb.MVCCMetadata // only set after LoadFromRawValue
+	Ent  raftpb.Entry          // must be pre-populated when calling Load
```
nit: consider typing this out as `Entry`, it reads better when it's embedded elsewhere.

It also stutters a bit to have to do `Entry.Entry` everywhere. Maybe we should embed this?

erikgrinaker commented on pkg/kv/kvserver/raftlog/entry.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// CC returns CC1 or CC2 as an interface, if set. Otherwise, returns nil.
+func (e *Entry) CC() raftpb.ConfChangeI {
```
nit: consider typing this out as `ConfChange`, ditto with the fields. It's too terse when encountered in the code.

erikgrinaker reviewed (commented):
Nice to see this cleanup!

sumeerbhola reviewed (commented):
drive-by comments. Thanks for doing this!

> Reviewed 8 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-:-MvJlw701CxSrLx1MDc0:b-q1yu4t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[pkg/kv/kvserver/raftlog/entry.go, line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvJiO2_78k-MNRzcNcs:-MvJiO2a3IbenLb1sii9:b3aqybt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/entry.go#L50)):*
> ```Go
> }
> 
> // Load populates the Entry from the Ent field. Will reset the Meta field.
> ```

wondering why this comment mentions the `Meta` field. If it populates `Entry` doesn't that imply that everything will be reset?
I guess it is because of this weird behavior where sometimes we have to unmarshal `MVCCMetadata` to get a `raftpb.Entry` and sometimes we are handed a `raftpb.Entry` directly. Is the latter case the one where we have it in the cache?
___
*[pkg/kv/kvserver/raftlog/entry.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVg-r1-98:-MvJjVwO5h6SrooLTzww:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/entry.go#L98)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: consider typing this out as `ConfChange`, ditto with the fields. It's too terse when encountered in the code.
</blockquote></details>

+1
___
*[pkg/kv/kvserver/raftlog/iterator.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r1-52:-MvJkxjyCJCeiAAHP_Lv:b-8e10la) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/iterator.go#L52)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Seek shouldn't take a bound. Put the iterator construction and bounds in `NewIterator`, and have `Seek` take a single index.
</blockquote></details>

And rename it SeekGE just to be consistent with other code.
___
*[pkg/kv/kvserver/raftlog/iterator.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r1-80:-MvJkKMOBByyuSetYwA9:b-xvgm8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/iterator.go#L80)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Seems iffy to do this as a side-effect of `Valid()`. This iterator pattern that we generally use is pretty awkward in terms of error handling and such (particularly decoding errors), but what we usually end up doing is to call into this decoding in `Seek` and `Next`, and set an error in `Iterator.err` which is surfaced from `Valid()`.

Making `UnsafeEntry()` return a decoding error might be the saner choice here -- I think that's what we want to do with e.g. `MVCCIterator` down the road too.
</blockquote></details>

MVCCIterator will probably move to a model where `Next` (and seeks) will return `(valid bool, err error)`, since the interface call overhead of calling `Valid` every time is not insignificant (there is a TODO in engine.go). We may as well adopt a similar approach here, which means we would decode in Next and SeekGE.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Addressed the comments, ready for another round. I still need to do a little perf work here (left a comment there as well).

> Dismissed @erikgrinaker and @sumeerbhola from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-:-MvdixFL6F8EXJVaiXuU:b-b25yxq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @sumeerbhola)

___
*[pkg/kv/kvserver/replica_application_cmd.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVi-r1-45:-MvcaUI6CMu87J4MiHRM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/replica_application_cmd.go#L45)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Remove the comment -- `ent` no longer exists.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_application_cmd.go, line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvdikHD8xQCnKgXw9pi:-MvdikHD8xQCnKgXw9pj:b3fspgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec57fc4247611bafd80d39a6f45912acdf369c3e/pkg/kv/kvserver/replica_application_cmd.go#L187)):*
> ```Go
> 
> 	var err error
> 	d.Entry, err = raftlog.MakeEntry(*e)
> ```

I need to look at this one again, I think as is I'm allocating more than before because `MakeEntry` will internally put the return value on the heap.
___
*[pkg/kv/kvserver/raftlog/iterator.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r1-52:-Mvcb24N85jZF2QSWA6W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/iterator.go#L52)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

And rename it SeekGE just to be consistent with other code.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/raftlog/iterator.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r1-80:-Mvc_yjM52omaWGlojeR:b2urb61) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56a307ab6ef1a706b19201f5fc4e029d48d6451/pkg/kv/kvserver/raftlog/iterator.go#L80)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

MVCCIterator will probably move to a model where `Next` (and seeks) will return `(valid bool, err error)`, since the interface call overhead of calling `Valid` every time is not insignificant (there is a TODO in engine.go). We may as well adopt a similar approach here, which means we would decode in Next and SeekGE.
</blockquote></details>

I've implemented Sumeer's suggestion.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
I spent a few hours figuring out why once I pushed the new code all packages were timing out. I haven't figured it out yet, but ultimately it seems to be caused by a rebase on master I carried out. I seem to be able to pass all tests when rebased on 4ea464d8376402e55bf85d4e69efd4abba33acdd (the initial commit this PR started out on top of) but on 18e41933ffbb0bdbdf7062623868fa0725bbc874 it's broken.

Since I need to figure this out anyway, no need to review again yet. I will likely end up making some changes anyway.

> Reviewed 2 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-:-MvearEX7BOztekafAfZ:bhr5yj2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

tbg commented:
git bisect start '--first-parent'
# good: [4ea464d8376402e55bf85d4e69efd4abba33acdd] Merge #76120
git bisect good 4ea464d8376402e55bf85d4e69efd4abba33acdd
# bad: [18e41933ffbb0bdbdf7062623868fa0725bbc874] Merge #76406
git bisect bad 18e41933ffbb0bdbdf7062623868fa0725bbc874
# bad: [432a383c1a7509762ce66a2b9ae7869d82da5f31] Merge #76242 #76270 #76272
git bisect bad 432a383c1a7509762ce66a2b9ae7869d82da5f31
# good: [8ec5bf1453d6ca9d19e6b587cc421827befead3d] Merge #75541 #75738 #75898 #75912 #75981 #75999 #76078 #76193 #76200 #76205
git bisect good 8ec5bf1453d6ca9d19e6b587cc421827befead3d
# bad: [4d95f56f5e47b99f4c7ec2a613d68cef4e004bc0] Merge #75889 #76209 #76236
git bisect bad 4d95f56f5e47b99f4c7ec2a613d68cef4e004bc0
# bad: [03c50b0dc6eb45f79ebb7e4ee677f1dafad69565] Merge #75905
git bisect bad 03c50b0dc6eb45f79ebb7e4ee677f1dafad69565
# bad: [a7b1c17c240c935056f87f29dfc1ee75611c371e] Merge #76095 #76163 #76166 #76188 #76191 #76192 #76194
git bisect bad a7b1c17c240c935056f87f29dfc1ee75611c371e
# good: [71e8cac55389008e9cef6393657ab39076f5dd92] Merge #76089 #76091 #76162
git bisect good 71e8cac55389008e9cef6393657ab39076f5dd92
# first bad commit: [a7b1c17c240c935056f87f29dfc1ee75611c371e] Merge #76095 #76163 #76166 #76188 #76191 #76192 #76194

tbg commented:
So basically it's https://github.com/cockroachdb/cockroach/pull/76095. I can work with that.

tbg commented:
Worked on this a bit today (thanks, Flex Friday!) and this should actually work now. Let's hold off on reviews until the stability period stuff is a bit less hectic, though.

tbg commented:
@pavelkalinnikov mind giving this a look next week? I think it's time to get this in now.

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,142 @@
+	"github.com/cockroachdb/errors"
+)
+
+type storageIterI interface {
```
nit: Would drop the "I" suffix.

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,142 @@
+	return true, nil
+}
+
+// SeekGE positions the Iterator at the first raft log with index greater than
```
Can there be gaps in the log indices? I.e. isn't the first log index >= idx equal to idx (or none)?

I imagine a gap can happen when the index is compacted away? Will this be an error then, or depends on who iterates?

Maybe would be worth elaborating nuances around entries with consecutive indices, somewhere in the `Iterator` comment.

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,142 @@
+	ctx context.Context,
+	rangeID roachpb.RangeID,
+	eng ReaderI,
+	lo, hi uint64, // lo is inclusive, hi exclusive
```
nit: Move or dup this comment in the method comment, so that it's reflected in docs/popups too.

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,142 @@
+// available via Entry. It returns (false, nil) if there are no more
+// entries.
+// Note that Iterator does not check for gaps in the log.
+func (it *Iterator) Next() (bool, error) {
```
Will iteration ever jump by more than +1 index? What precisely does "next entry" mean?

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,142 @@
+		return err
+	}
+	for ; ok; ok, err = it.Next() {
+		ent := it.Entry()
```
Is reading one extra entry unavoidable here? E.g. could the storage iterator have an upper bound that it more efficiently terminates at?

pav-kv reviewed (commented):
Looks good, ta! Left a few nits, and questions about the semantics.

I think this PR could have been easier to review if split so that one commit introduced the `Entry` type, and another one for the `Iterator`.

tbg reviewed (commented):
> I think this PR could have been easier to review if split so that one commit introduced the Entry type, and another one for the Iterator.

Thanks for the feedback, I'll be a bit more aggressive about splitting next time. I find that it does slow me down quite a bit so it needs to be worth it.

I'd like to add some more unit testing to this so this needs some more work, but please PTAL at the fixup commits. Additional testing will come in follow-up commits (I don't want to merge before we have this), so by establishing a checkpoint here you don't have to revisit it later.

> Reviewed 7 of 7 files at r9.
> Dismissed @erikgrinaker from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-:-NGWkxTOAmd_Xg6PxOQt:b-v4qsux)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @pavelkalinnikov, and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_application_cmd.go` line 45 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVi-r10-45:-NGWq4wICDkgQXc4KZJy:br2j6tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/replica_application_cmd.go#L45)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

How about removing the `decodedRaftEntry` wrapper, and directly inlining `entry *raflog.Entry`? There seems to be no reason to keep it?
</blockquote></details>

Well spotted, done!
___
*[`pkg/kv/kvserver/replica_application_cmd.go` line 84 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVi-r10-84:-NGWgGh02VoxTFjlk_Fg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/replica_application_cmd.go#L84)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`ctx` seems unused all the way down the stack from here
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/raftlog/entry.go` line 86 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVg-r10-86:-NGWgKbmAol5cMOfgFGR:b-s3oie9) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/raftlog/entry.go#L86)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`ConfChangeV1`
</blockquote></details>

The type is `raftpb.ConfChange`, and that's what we're unmarshalling (into).
___
*[`pkg/kv/kvserver/raftlog/iterator.go` line 89 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r10-89:-NGWgjJw3kdew3241PMi:b-kjdatq) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/raftlog/iterator.go#L89)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Can there be gaps in the log indices? I.e. isn't the first log index >= idx equal to idx (or none)?

I imagine a gap can happen when the index is compacted away? Will this be an error then, or depends on who iterates?

Maybe would be worth elaborating nuances around entries with consecutive indices, somewhere in the `Iterator` comment.
</blockquote></details>

There are never any gaps. I'll make this clear on the Iterator comment.
___
*[`pkg/kv/kvserver/raftlog/iterator.go` line 101 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r10-101:-NGWh7fI8GGViBYm1P3G:b-g718l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/raftlog/iterator.go#L101)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Will iteration ever jump by more than +1 index? What precisely does "next entry" mean?
</blockquote></details>

Index+1. I added a comment.
___
*[`pkg/kv/kvserver/raftlog/iterator.go` line 130 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76126#-MvF_CeqL929rC68oqVh-r10-130:-NGWiMjgAptMdYRqO78l:bdg1to) ([raw file](https://github.com/cockroachdb/cockroach/blob/25582ad5088554c68105ec8f79b2db788d0aeeeb/pkg/kv/kvserver/raftlog/iterator.go#L130)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Is reading one extra entry unavoidable here? E.g. could the storage iterator have an upper bound that it more efficiently terminates at?
</blockquote></details>

Yes, see the TODO in `NewIterator`. Actually I went ahead and implemented it.


<!-- Sent from Reviewable.io -->

tbg committed:
raftlog: add BenchmarkNewEntry

Protobuf makes it hard to do well here.

```
BenchmarkNewEntry/fromRawValue=false,release=false-10   1319 ns/op   4890 B/op  10 allocs/op
BenchmarkNewEntry/fromRawValue=false,release=true-10    1128 ns/op   4442 B/op   9 allocs/op
BenchmarkNewEntry/fromRawValue=true,release=false-10    2611 ns/op  13213 B/op  13 allocs/op
BenchmarkNewEntry/fromRawValue=true,release=true-10     2536 ns/op  12766 B/op  12 allocs/op
```

Release note: None

tbg commented:
Added another commit with unit testing, which can also be reviewed.

pavelkalinnikov commented on pkg/kv/kvserver/raftlog/iterator.go:
```diff
@@ -0,0 +1,168 @@
+type IterOptions struct {
+	// Hi ensures the Iterator never seeks to any Entry with index >= Hi. This is
+	// useful when the caller is interested in a slice [Lo, Hi) of the raft log.
+	Hi uint64
```
This probably is a result of `MVCCIterator` interface, but it's a bit asymmetrical that the iterator has to be seeked to the `Lo` position, whereas the `Hi` position can be specified at the creation time.

Would it make sense to have an API where you set both at creation?

Can you and do you ever need to re-`Seek` an iterator to a lower position? Higher position? Is there a benefit (performance?) in reusing iterators?

If there is low benefit in reusing, I could imagine a simpler / more restictive API:
- `NewMVCCIterator` / `NewIterator` return an iterator with defined `[Lo,Hi)` straight away.
- It's a "forward" iterator, so only has `Next`, doesn't have `Seek`, or allows only seeking forward.

pavelkalinnikov replied:
Given that the only user of this is `Visit`, the above should make sense (side note: could even unexport it for now?). Are there other potential users of this iterator? Is the common denominator still the scan-once pattern that `Visit` does?

tbg replied:
There was an earlier iteration of the Iterator as well, the current form is after suggestions from the storage team. I think we can iterate on this more in the future but I'd like to get this PR merged if there are no real objections.

tbg replied:
As per a DM chat, left TODO comment on Iterator.

pavelkalinnikov commented on pkg/kv/kvserver/replica_application_cmd.go:
```diff
@@ -44,8 +42,7 @@ import (
 type replicatedCmd struct {
-	ent              *raftpb.Entry // the raft.Entry being applied
-	decodedRaftEntry               // decoded from ent
+	*raftlog.Entry
```
There is a `raftentry` package. Not sure it's about the same thing, but maybe there should be some integration with it in a follow-up to this PR.

tbg replied:
I have a hard time waterfalling the eventual package structure, so for now I'm happy just making any refactor that squarely moves us towards a better organization of the codebase. `raftentry` currently houses the `raftentry.Cache` and it makes sense that it will ultimately somehow live closer to the raft log.
I'd like to spend some time pulling more components out of `kvserver` to unblock the pending work though, before focusing on how all of these leaf packages should be organized.

tbg replied:
Left a TODO comment as discussed in DM.

tbg commented:
TFTR!

bors r=pavelkalinnikov
